### ISSUE TITLE
exp: RL post-training on contacts-v1 with a dense per-contact reward

### ISSUE THREAD

**Issue by timodonnell** [Open-Athena/MarinFold#208]
RL post-training on the **base `<contacts-v1>` format only**, with a dense per-contact reward. This is the simpler precursor to [#200](https://github.com/Open-Athena/MarinFold/issues/200), which ran the same reward on the multi-draft format and could not separate candidate quality from candidate diversity.

## Why this before more multi-draft work

#200's reward provably worked and its headline metric provably didn't move. Paired over 554 eval proteins: per-contact precision **+0.0085 (+4.6σ)** and first-candidate F1 **+0.0128 (+5.1σ)**, while best-of-N F1 came out **+0.0008 (+0.4σ)** because sections fell 1.35 (−7.5σ) and inter-section Jaccard rose 0.0087 (+7.7σ). best-of-N is quality × spread and the two cancelled.

Dropping multi-draft removes the spread-across-sections axis entirely. What remains is answerable: **does a dense per-contact reward improve single-shot contacts-v1 prediction?** Two further reasons to do it here:

- `first_f1` — the metric that moved most in #200 — is generated with no prior candidates in context, so it is the closest analogue to single-shot output. The multi-draft machinery diluted the gain.
- Plain generation is ~4× cheaper: measured in #200, 37 s per sampling call against 162 s for multi, with an 860-token budget against 5,344.

This would also be MarinFold's first RL on the base task. #98/#100/#120/#175 are all SFT-flavoured; #120's negative result was imitation of filtered samples, not a gradient on individual contact decisions.

## Starting model

**`contacts-v1-exp199-1.5B`** — now MarinFold's default, landed in [#207](https://github.com/Open-Athena/MarinFold/pull/207).

- `marinfold/marinfold/MODELS.yaml` → `hf://buckets/open-athena/MarinFold/checkpoints/prot-exp199-cw-cv1-s02-m1-p06-aug/hf/step-145199`
- Qwen3 1.47B, trained **from scratch** on 16 H100 nodes (cw-rno2a), WSD, 145,200 steps, lr 1e-3 / wd 0.2 / global batch 128, 152.3B tokens over a 50/50 AFDB + ESM-Atlas mixture with amino-acid augmentation. Same architecture as #163's model, so #200's `MODEL_CONFIG` carries over unchanged.

Deliberately **not** #163 arm F: that is the multi-draft model, and this experiment is about the base format.

**Prerequisite: republish as an HF model repo.** The registry URL is a *bucket* path, and the RL rollout worker resolves its tokenizer through `levanter.tokenizers.load_tokenizer`, which accepts a local directory, a `mirror://` ref, or an HF **repo id** — not a bucket path or a `gs://` URL. vLLM streams weights from object storage happily, so the failure lands only on the tokenizer, inside a worker, after the gang has scheduled. Reuse #200's `publish_checkpoint_hf.py`, which also preflights `rope_theta` and the tokenizer before uploading.

Note the rope hazard is already handled for this checkpoint: MarinFold's republished copy states rope in both transformers 4.x and 5.x terms, because Eric's original export gave only a `rope_parameters` block, which 4.x silently ignores in favour of rope_theta 10000 where the model was trained at 500000 — worth 0.76 nats/token. Preflight anyway; do not assume.

**Distribution mismatch, accepted deliberately.** exp199 was trained on a 50/50 AFDB + ESM-Atlas mixture; the RL prompt pool here is AFDB **round-0 only** (#200's pool: 10,000 proteins, mean pLDDT 89.0, zero exact sequence overlap with the eval set). That is on purpose — ESM-Atlas ground truth is ESMFold2-distilled, and under a per-contact reward every wrong label is a directly mis-signed gradient on a specific token, which is worse than merely noisy. Worth watching for drift on the ESM-Atlas half of the distribution, but label quality wins here.

## The reward

Two terms, both landing per token, on `<contacts-v1>` prompts only (mode token id 2). Reuses #200's `contact_rewards.py`, which already supports `mode="plain"`.

**1. Stepwise, per emitted contact.** For each `<contact> <pI> <pJ>` triple, with `x = 1` iff the pair is in ground truth (seq-index, sep ≥ 6, not already emitted):

```
r_step = +(1 - p̄)          if x == 1
       = −p̄ · δ^e          if x == 0     (e = earlier errors in the section)
```

spread over the triple's three tokens. `p̄` is the policy's own EMA per-contact precision, so `E[r_step] ≈ 0` at current performance and the gradient says *beat yourself*. This matters more than any other constant: precision is ~0.23, so a **fixed** penalty makes "emit nothing" optimal. `δ < 1` reflects that a wrong contact may be the consequence of an earlier one.

Penalising wrong contacts is also what preserves the incentive to **end a rollout early when the model has nothing good left** — which is wanted, not merely tolerated.

**2. Document-level: leave-one-out marginal contribution to consensus.** Not the single rollout's F1. Score the group's consensus vote with rollout *i* included versus excluded; the difference is *i*'s advantage.

This is the load-bearing design choice, and it is there because of a measured prior result rather than a hunch. The reported metric is a **consensus over 100 rollouts** (`score_rollout_worker.py`: *"n resampled rollouts per protein, voted into a dense vote matrix"*, top-L by vote with pairwise tie-break). #82's own README records that **"over-sharpening collapses the vote"** — sharpening the sampling distribution *hurts*, and T=1.0/p=0.95 is near-optimal. So a precision-only reward can raise per-rollout precision and **lower the reported metric** by making all 100 rollouts agree. That is #200's failure arriving through a different door.

Scoring marginal consensus contribution makes the objective literally the deployed metric, and rewards a rollout for contributing something its siblings didn't — so diversity is defended by construction rather than by a hyperparameter. It also reuses RLOO's existing leave-one-out group structure, so it is a small change to #200's `dense_loss.py`.

`A_t = λ_step · r_step(t) + λ_doc · A_consensus`. **The λ ratio is the primary axis** — that is the lever #200 identified.

## Eval

- **Primary — consensus R-precision at n=100**, all and long bands, via #82's `score_rollout_worker.py` (`--n-rollouts 100 --top-k -1 --top-p 0.95`). This is the number reported in model summaries. **Baseline to beat, from #204: 0.587 (all) / 0.542 (long)**, 0.666 (short) / 0.636 (medium) — re-measure it in the same batch rather than quoting it, so the comparison is paired. For scale, the progression is #117 0.534 → #166 0.562 → #199 0.587.

  Be clear-eyed about the bar: 0.587 is already statistically indistinguishable from single-sequence Protenix-v2's 0.603 (paired −0.016, 95% CI [−0.041, +0.009]). This is the strongest starting point MarinFold has, so a gain is more meaningful and less likely than it would have been from #117. A null result against this baseline says less about the method than a null against a weaker model would. ~554 × 100 rollouts ≈ 15.7M tokens, roughly 15 min on a v5p-8 without logprobs.
- **Secondary — single-rollout precision, recall and F1, reported separately.** The main risk is a precision/recall trade presented as an improvement, and only separate reporting catches it.
- **Monitored, as early warning for vote collapse** — policy entropy (#200 logged 2.618) and cross-rollout agreement. If entropy falls while per-rollout precision rises, the consensus is being sharpened away.
- **Output length as an outcome, not a guardrail.** Some length reduction is *wanted* here. State an acceptable range up front, because the stop-incentive and collapse-to-silence are the same mechanism at different strengths, separated only by the α ratio.
- **Cheap check — pairwise teacher-forced R-precision** (#163's `rprec_worker_tpu.py`). Secondary only: it is computed from the probability matrix, not from rollouts, so RL on rollouts may not move it at all.

## Things that will bite

Carried from #200, where each cost real time:

- **Pin marin at `0.2.76.dev31155643335`.** 0.2.77 (2026-08-08) is the first release without `marin.rl`, which is `ContactsDenseLoss`'s base class and the whole environment API. The pin cannot be advanced; marin's direction is now SkyRL, which is a rewrite.
- **Co-locate data with compute.** Compute in us-central1 with data in us-east5 dies with `TransferBudgetExceeded` about an hour in, once rollout spill and checkpoints start flowing. Prompt *reads* are trivially small; the spill is not. #200's `check_region_locality` refuses it at config time.
- **A run cannot terminate itself.** The trainer's `mark_completed` RPC has no timeout and the rollout worker swallows poll errors, so nothing reaps a finished trainer. Use #200's `reap.py`, which detects completion through W&B.
- **`canonical_model_name` must be a registered `MODEL_MAPPINGS` key** (`Qwen/Qwen3-1.7B`), not a descriptive string — it feeds both a renderer substring match and an exact-key lookup on the weight-transfer path.
- **The engine needs an HF repo id**, not a `gs://` path: `load_tokenizer` rejects URLs even though vLLM streams weights from GCS happily.
- **Launch arms as independent jobs and checkpoint on a step interval.** #200 lost two of three arms because one driver preemption stranded its siblings, and their 20-minute rolling checkpoints lagged training by ~30 steps, leaving nothing clean to evaluate.
- **v5p-16 had 0 ready slices** while v5p-8 had 121. Size the trainer to what schedules, and remember marin's DAPO normalisation divides by batch size, so learning rates do not transfer across a batch-size change.

## Scope

Mostly configuration on top of #200's code: `contact_rewards.py` already takes `mode="plain"`, `contacts_env.py` takes mode as a parameter, and the curriculum collapses to one lesson. New work is the consensus-marginal document term and the eval swap to `score_rollout_worker.py`. One arm at 150 steps should run well under an hour on 1 + 4 v5p-8.

**Comment by timodonnell** [comment]
Unblocked — `contacts-v1-exp199-1.5B` landed in #207 and is now the default. Issue body updated with the registry path, the model's provenance, and three consequences worth calling out:

**The baseline moved up.** #204 puts it at **0.587 (all) / 0.542 (long)**, not the ~0.53–0.56 I had used for scale. That is the strongest starting point MarinFold has, and it is already statistically indistinguishable from single-sequence Protenix-v2 (paired −0.016, 95% CI [−0.041, +0.009]). Worth stating before we run: a null result against this baseline says less about the method than a null against #117 would have.

**A prerequisite appears.** The registry URL is a *bucket* path, and the RL rollout worker needs an HF **repo id** — `load_tokenizer` takes a local dir, a `mirror://` ref, or a repo id, and nothing else. vLLM streams weights from object storage fine, so the failure would land only on the tokenizer, inside a worker, after the gang scheduled. #200's `publish_checkpoint_hf.py` handles it and preflights rope + tokenizer on the way out.

**One design note.** exp199 trained on a 50/50 AFDB + ESM-Atlas mixture; the RL pool is AFDB round-0 only. That is deliberate — ESM-Atlas ground truth is ESMFold2-distilled, and under a per-contact reward a wrong label is a directly mis-signed gradient on a specific token rather than just noise. Label quality wins, but it is a real distribution mismatch and we should watch for drift on the ESM-Atlas half.
